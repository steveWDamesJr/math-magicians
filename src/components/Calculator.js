import React, { useState } from 'react';
import calculate from '../logic/calculate';

function Calculator() {
  const [state, setState] = useState({ total: '0', next: null, operation: null });

  const handleClick = (e) => {
    const buttonContents = e.target.textContent;
    setState((prevState) => ({
      ...prevState,
      total: buttonContents,
    }));
    const result = calculate(state, buttonContents);
    setState(result);
  };

  const { total, next, operation } = state;
  const onClick = (e) => { handleClick(e); };

  return (
    <main className=" w-screen h-screen flex justify-between">
      <h2>Let&apos;s do some math!</h2>
      <div className="cal-container flex">
        <div className="calculator">
          <div className="w-96 h-auto bg-white rounded-2xl shadow-xl">
            <div className="w-auto h-auto">
              <div className="pr-1 btn-zinc col-span-4 text-right cal-screen" data-testid="cal-screen-out">{total || next ? `${total || ''} ${operation || ''} ${next || ''}` : 0}</div>
              <div className="flex justify-between">
                <button type="button" className="btn-grey" key="first" onClick={onClick}>AC</button>
                <button type="button" className="btn-grey" key="second" onClick={onClick}>+/-</button>
                <button type="button" className="btn-grey" key="third" onClick={onClick}>%</button>
                <button type="button" className="btn-orange" key="fourth" onClick={onClick}>รท</button>
              </div>
              <div className="flex justify-between">
                <button type="button" className="btn-grey" key="fifth" onClick={onClick}>7</button>
                <button type="button" className="btn-grey" key="sixth" onClick={onClick}>8</button>
                <button type="button" className="btn-grey" key="seventh" onClick={onClick}>9</button>
                <button type="button" className="btn-orange" key="eight" onClick={onClick}>x</button>
              </div>
              <div className="flex justify-between">
                <button type="button" className="btn-grey" key="ninth" onClick={onClick}>4</button>
                <button type="button" className="btn-grey" key="tenth" onClick={onClick}>5</button>
                <button type="button" className="btn-grey" key="eleventh" onClick={onClick}>6</button>
                <button type="button" className="btn-orange" key="twelfth" onClick={onClick}>-</button>
              </div>
              <div className="flex justify-between">
                <button type="button" className="btn-grey" key="thirteenth" onClick={onClick}>1</button>
                <button type="button" className="btn-grey" key="fourteenth" onClick={onClick}>2</button>
                <button type="button" className="btn-grey" key="sixteenth" onClick={onClick}>3</button>
                <button type="button" className="btn-orange" key="first" onClick={onClick}>+</button>
              </div>
              <div className="flex justify-between last">
                <button type="button" className="w-64 zero" key="seventeenth" onClick={onClick}>0</button>
                <button type="button" className="w-32 btn-grey point" key="eighteenth" onClick={onClick}>.</button>
                <button type="button" className="w-32 btn-grey equals" key="nineteenth" onClick={onClick}>=</button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  );
}

export default Calculator;
