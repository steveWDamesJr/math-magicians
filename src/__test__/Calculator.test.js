import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Calculator from '../components/Calculator';
import '@testing-library/jest-dom';

it('renders correctly', () => {
  const tree = renderer.create(<Calculator />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('Perform calculation operations', () => {
  test('Click one  number to display on screen', () => {
    render(<Calculator />);
    const result = screen.getByTestId('cal-screen-out');
    userEvent.click(screen.getByText('7'));
    expect(result.innerHTML).toMatch('7');
  });
  test('Click equals button to calculate and render result of operation on screen', () => {
    render(<Calculator />);
    const result = screen.getByTestId('cal-screen-out');
    userEvent.click(screen.getByText('2'));
    userEvent.click(screen.getByText('4'));
    userEvent.click(screen.getByText('+'));
    userEvent.click(screen.getByText('4'));
    userEvent.click(screen.getByText('='));
    expect(result.innerHTML).toMatch('28');
  });
  test('Click two numbers to display output on screen', () => {
    render(<Calculator />);
    const result = screen.getByTestId('cal-screen-out');
    userEvent.click(screen.getByText('5'));
    userEvent.click(screen.getByText('2'));
    expect(result.innerHTML).toMatch('52');
  });

  test('Click modulus button to see remainder of two numbers on screen', () => {
    render(<Calculator />);
    const result = screen.getByTestId('cal-screen-out');
    userEvent.click(screen.getByText('AC'));
    userEvent.click(screen.getByText('4'));
    userEvent.click(screen.getByText('0'));
    userEvent.click(screen.getByText('%'));
    userEvent.click(screen.getByText('2'));
    expect(result.innerHTML).toMatch('0');
  });
});
