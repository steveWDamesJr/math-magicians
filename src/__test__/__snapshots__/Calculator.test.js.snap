// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className=" w-screen h-screen flex justify-between"
>
  <h2>
    Let's do some math!
  </h2>
  <div
    className="cal-container flex"
  >
    <div
      className="calculator"
    >
      <div
        className="w-96 h-auto bg-white rounded-2xl shadow-xl"
      >
        <div
          className="w-auto h-auto"
        >
          <div
            className="pr-1 btn-zinc col-span-4 text-right cal-screen"
            data-testid="cal-screen-out"
          >
            0  
          </div>
          <div
            className="flex justify-between"
          >
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              AC
            </button>
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              +/-
            </button>
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              %
            </button>
            <button
              className="btn-orange"
              onClick={[Function]}
              type="button"
            >
              รท
            </button>
          </div>
          <div
            className="flex justify-between"
          >
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              7
            </button>
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              8
            </button>
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              9
            </button>
            <button
              className="btn-orange"
              onClick={[Function]}
              type="button"
            >
              x
            </button>
          </div>
          <div
            className="flex justify-between"
          >
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              4
            </button>
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              5
            </button>
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              6
            </button>
            <button
              className="btn-orange"
              onClick={[Function]}
              type="button"
            >
              -
            </button>
          </div>
          <div
            className="flex justify-between"
          >
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              1
            </button>
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              2
            </button>
            <button
              className="btn-grey"
              onClick={[Function]}
              type="button"
            >
              3
            </button>
            <button
              className="btn-orange"
              onClick={[Function]}
              type="button"
            >
              +
            </button>
          </div>
          <div
            className="flex justify-between last"
          >
            <button
              className="w-64 zero"
              onClick={[Function]}
              type="button"
            >
              0
            </button>
            <button
              className="w-32 btn-grey point"
              onClick={[Function]}
              type="button"
            >
              .
            </button>
            <button
              className="w-32 btn-grey equals"
              onClick={[Function]}
              type="button"
            >
              =
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</main>
`;
